### PR TITLE
[Registers] Setup registers for AArch64

### PR DESCRIPTION
**Pull request by dongju-chae**
This PR makes setup for AArch64 registers in the init function.

**changelog:**
- V2: Add the enum for Arm64 registers.

Signed-off-by: Dongju Chae <dongju.chae@samsung.com>

### CODE REVIEW

File: orc/orcarm.h [hunk 1]
```diff
@@ -24,7 +24,23 @@ typedef enum {
...
+  ORC_ARM_R22,
+  ORC_ARM_R23,
+  ORC_ARM_R24,
+  ORC_ARM_R25,
+  ORC_ARM_R26,
+  ORC_ARM_R27,
+  ORC_ARM_R28,
+  ORC_ARM_R29,
+  ORC_ARM_R30,
+  ORC_ARM_R31
```
Comment (myungjoo) [inline orc/orcarm.h]: Is this really ok? Isn't this required to have the correct order from reg-base?

Comment (dongju-chae) [reply orc/orcarm.h]: Yes, I think it's okay. It seems that the current ORC imp. for other architectures use the overlapped registers for both 32bit and 64bit.
For example, in `orc/orcx86.h`, we can see like
```
enum {
  X86_EAX = ORC_GP_REG_BASE,
  X86_ECX,
  X86_EDX,
  X86_EBX,
  X86_ESP,
  X86_EBP,
  X86_ESI,
  X86_EDI,
  X86_R8,
  X86_R9,
  X86_R10,
  X86_R11,
  X86_R12,
  X86_R13,
  X86_R14,
  X86_R15
};
```
From `x86_R8`, there are x86_64 specific registers.

Comment (myungjoo) [reply orc/orcarm.h]: image

No. ```orc/orcx86.h``` has the exact ordering with register numbers from the base.

Comment (dongju-chae) [reply orc/orcarm.h]: Yes, I think x86 have confusing synonyms (e.g., EAX, ECX) which are not well-matched with RX.
But, In AAarch32/64, they have relatively clear register names starting from R0.
Of course, they also have synonyms but according to the following figures, I think it's okay to define them as suggested.
aarch32
aarch64

In the case of generic-purpose registers (i.e., R0-R30), I think it's okay. 
But, my concern is vector registers in AArch64, which are surely different from AArch32.
```
Unlike in AArch32, in AArch64 the 128-bit and 64-bit views of a SIMD and 
Floating-Point register do not overlap multiple registers in a narrower view, 
so q1, d1 and s1 all refer to the same entry in the register bank.
```

File: orc/orcprogram-neon.c [hunk 2]
```diff
@@ -150,36 +157,74 @@ orc_compiler_neon_init (OrcCompiler *compiler)
...
+    }
+  } else {
+    /** AArch32
+     * 16 32-bit generic-purpose registers (R0-R15)
+     * 32 64-bit vector registers (smaller registers are packed into larger ones)
+     */
+    for(i=ORC_GP_REG_BASE;i<ORC_GP_REG_BASE+16;i++){
+      compiler->valid_regs[i] = 1;
+    }
+    for(i=ORC_VEC_REG_BASE+0;i<ORC_VEC_REG_BASE+32;i+=2){
```
Comment (dongju-chae) [inline orc/orcprogram-neon.c:199]: That's why here, existing setup for vector registers (aarch32) use `i+=2` as a counter

**Comment by dongju-chae** [comment]
@myungjoo I understand your concerns. If the compiler directly uses these macros defined in `orcarm.h` for compilation, it could be a problem because, for example, a link register (LR) in 32bit is R14 but 64bit uses R30 as LR.

However, the orc compiler uses common register maps such as `valid_regs, save_regs, used_regs, alloc_regs`, regardless of the architecture. Architecture-specific register macros are used only in initialization time, `orcprogram-neon.c`, to assign valid generic-purpose/vector registers. Special registers such as SP,FP, and LR, are marked `valid_regs = 0` to make them not-available for automatic register mapping.

**Review by wooksong — APPROVED** [review]
LGTM

**Review by myungjoo — APPROVED** [review]
LGTM. Ok, let's go ahead.

**Comment by dongju-chae** [comment]
@myungjoo @wooksong 
Sorry for the wrong information. I think it's better to declare new register macro for ARM64 @myungjoo said because I found that the actual compilation directly uses some register macros (e.g., `ORC_VAR_IP`).

Of course, as ORC works with the enum, we can use the existing macros but it will probably make some confusions. I will add a enum named `ORC_ARM64_RXX`.

**Review by myungjoo — APPROVED** [review]
LGTM

**Pull request merged**